pcf/test/particle/aws/test_particles.py

```diff
@@ -0,0 +1,54 @@
+import moto
+import json
+import pytest
+import boto3
+import os
+import placebo
+from pcf.core import State
+from pcf.core import particle_flavor_scanner
+from contextlib import ExitStack
+
+from pcf.particle.aws.route53 import hosted_zone
+from pcf.particle.aws.sqs.sqs_queue import SQSQueue
+from pcf.particle.aws.cloudfront.cloudfront import CloudFront
+
+with open("testdata.json") as f:
+    testdata = json.load(f)
+
+values = testdata.values()
+values = [tuple(v) for v in values]
+
+
+@pytest.mark.parametrize("definition,updated_definition,test_tag", values, ids=list(testdata.keys()))
+def test_apply(definition, updated_definition, test_tag):
+    flavor = definition.get("flavor")
+    particle_class = particle_flavor_scanner.get_particle_flavor(flavor)
+    session = None
+    with ExitStack() as stack:
+        if test_tag[0] == "placebo":
+            session = boto3.Session()
+            dirname = os.path.dirname(__file__)
+            filename = os.path.join(dirname, test_tag[1])
+            pill = placebo.attach(session, data_path=filename)
+            pill.playback()
+        else:
+            for context in test_tag:
+                stack.enter_context(getattr(moto, context)())
+
+        particle = particle_class(definition, session)
+        particle.set_desired_state(State.running)
+        particle.apply()
+
+        assert particle.get_state() == State.running
+
+        if updated_definition:
+            particle = particle_class(updated_definition)
+            particle.set_desired_state(State.running)
+            particle.apply()
+
+            assert particle.is_state_definition_equivalent()
```

Comment:
i think we need to assert current def = input def.

Reply:
isn't that what is_state_definition_equivalent() is doing?